### PR TITLE
Add support for arm64 build (tested on Mac M1)

### PR DESCRIPTION
**Pull request by PCatinean**
This PR introduces support for ARM builds and has been tested on Mac M1, a basic rundown on the changes:

1) Download resources from the same release location but use the targeted architecture
2) Update GEOIP_UPDATER_VERSION to the latest version as arm64 builds are also available
3) Force wkhtmltopdf version to 0.12.6 since it also has an arm64 build (fail build otherwise)
4) Automatically upgrade odoo requirements.txt for arm64 build and force gevent==21.12.0 and greenlet==1.1.0 (minimum versions required for a successful build otherwise cython will thrown an error (i.e rc/gevent/_gevent_cgreenlet.pxd:181:6: Variables cannot be declared with 'cpdef'. Use 'cdef' instead.)

### CODE REVIEW

**Comment by ap-wtioit** [comment]
this should fix / implements #567 and #342 for version 16.0

**Comment by bosd** [comment]
Thanks for this PR, Could you also make one for V14?

**Comment by pedrobaeza** [comment]
Please rebase and solve conflicts, and do the pre-commit fixes in the same commit that introduces the error.

**Comment by PCatinean** [comment]
PR rebased and I also added arm builds for Odoo version 14, 15 and 17. I have build tested  all of them locally on Mac M1

**Comment by pedrobaeza** [comment]
Thanks for it. Seems OK. The only point I don't know right now is where to indicate that the pushed build is for both platforms. Do you know it?

**Comment by PCatinean** [comment]
I think we need to adjust the ./hooks/build script or use the available github actions but either way it seems docker buildx is required for multi-platform builds: https://docs.docker.com/build/ci/github-actions/multi-platform.

I will make another branch on my fork and link it to my docker hub then do some tests there and later propose those changes to this PR as well.

In the meantime people can use this PR to build the *-onbuild image locally.

File: 14.0.Dockerfile [hunk 1]
```diff
@@ -35,8 +37,25 @@ ENV DB_FILTER=.* \
 # See https://github.com/$ODOO_SOURCE/blob/$ODOO_VERSION/debian/control
 RUN apt-get -qq update \
     && apt-get install -yqq --no-install-recommends \
-        curl \
-    && curl -SLo wkhtmltox.deb https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/${WKHTMLTOPDF_VERSION}/wkhtmltox_${WKHTMLTOPDF_VERSION}-1.buster_amd64.deb \
+        curl; \
+    if [ "$TARGETARCH" = "arm64" ]; then \
+        if [ "$WKHTMLTOPDF_VERSION" != "0.12.6.1" ]; then \
```
Comment (bosd) [inline 14.0.Dockerfile:42]: How is this supposed to work? As in line 7 the version is hard coded.
`ARG WKHTMLTOPDF_VERSION=0.12.5`

**Comment by bosd** [comment]
Would like to functional test this,  I have an arm server available.
But I'm light on experience. 

How can this be tested locally? :blush: :angel:
I've tried copying the contents of the 14.0-dockerfile into my `odoo/Dockerfile` of the [doodba-copier-template](https://github.com/Tecnativa/doodba-copier-template) but produces an error.

<details>

```

 > [base  8/10] COPY entrypoint.d common/entrypoint.d:
------
Dockerfile:90
--------------------
  88 |     COPY build.d common/build.d
  89 |     COPY conf.d common/conf.d
  90 | >>> COPY entrypoint.d common/entrypoint.d
  91 |     RUN mkdir -p auto/addons auto/geoip custom/src/private \
  92 |         && ln /usr/local/bin/direxec common/entrypoint \
--------------------
ERROR: failed to solve: failed to compute cache key: failed to calculate checksum of ref c65c21a4-8130-4ed7-9f2c-00eafd782e20::tw9or30qgl9vfkb72ljrcqszc: "/entrypoint.d": not found

```

</details>

**Comment by PCatinean** [comment]
@bosd I am not sure what is your setup so but here are a few steps on how to do this, if you need more details I think it's better to open a github discussion rather than continuing on this PR.

1) First you need to clone my branch locally `git clone https://github.com/pledra/doodba.git -b arm-builds doodba-arm`
2) Here you use either an ARM computer when building and just run something like `docker build -f 14.0.Dockerfile . -t doodba:14.0-arm-onbuild` or if you are not on ARM then you need to use docker buildx https://docs.docker.com/engine/reference/commandline/buildx/ and specify arm64 platform when building
3) Follow the steps at https://github.com/Tecnativa/doodba-copier-template to create a new template
4) After you have finished all the steps go into scaffolding/odoo/Dockerfile and replace `FROM ghcr.io/tecnativa/doodba:${ODOO_VERSION}-onbuild` with the above built image `doodba:14.0-arm-onbuild`
5) Now when you build the image locally from the scaffolding it should use the arm base image. 

One last thing to note and this will be probably a PR on other repositories, for the development environment some support images are missing arm builds as well.

**Comment by bosd** [comment]
@PCatinean Thanks!!
I managed to build it :tada: 
Had to mod the 14.0 and 15.0 dockerfiles to `WKHTMLTOPDF_VERSION=0.12.6.1` 

> One last thing to note and this will be probably a PR on other repositories, for the development environment some support images are missing arm builds as well.

Indeed on the dev env I got an error `This error is because there is no image build for linux/arm64/v8 platform.` on pgweb and the proxy.
For the proxys whitelist there is a pr in: https://github.com/Tecnativa/docker-whitelist/pull/15
I'm a bit puzzled on the pgweb one, as pgweb [supports arm](https://github.com/sosedoff/pgweb/wiki/Installation),

**Comment by PCatinean** [comment]
@bosd happy to hear it worked. Yes I will have to take a look at the other images as well, for the wkhtmltopdf version it is a build argument so you can just do --build-arg WKTHTMLTOPDF_VERSION=version and it will use that no need to change the original file

**Comment by bosd** [comment]
Yeah, We're getting there to arm support..
But there is still some todo on devel env.: after `inv start`
`The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested`
But that isunrelated to this pr :wink: 

<details>

```
root@debian:~/doodarm# inv start
WARN[0000] mount of type `volume` should not define `bind` option 
WARN[0000] mount of type `volume` should not define `bind` option 
[+] Running 4/0
 ✔ Volume "doodarm_filestore"                                                                                                                        Created0.0s 
 ✔ Volume "doodarm_db"                                                                                                                               Created0.0s 
[+] Running 5/5odarm-db-1                                                                                                                            Crea ✔ Volume "doodarm_filestore"                                                                                                                        Created0.0s e requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s ✔ Volume "doodarm_db"                                                                                                                               Created0.0s iner doodarm-odoo-1                                                                                                                          Crea[+] Running 5/5odarm-db-1                                                                                                                            Crea ✔ Volume "doodarm_filestore"                                                                                                                        Created0.0s e requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s ✔ Volume "doodarm_db"                                                                                                                               Created0.0s iner doodarm-odoo-1                                                                                                                          Crea[+] Running 5/5odarm-db-1                                                                                                                            Crea ✔ Volume "doodarm_filestore"                                                                                                                        Created0.0s e requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s ✔ Volume "doodarm_db"                                                                                                                               Created0.0s iner doodarm-odoo-1                                                                                                                          Crea[+] Running 5/5odarm-db-1                                                                                                                            Crea ✔ Volume "doodarm_filestore"                                                                                                                        Created0.0s e requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s ✔ Volume "doodarm_db"                                                                                                                               Created0.0s iner doodarm-odoo-1                                                                                                                          Crea[+] Running 5/5odarm-db-1                                                                                                                            Star ✔ Volume "doodarm_filestore"                                                                                                                        Created0.0s e requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s ✔ Volume "doodarm_db"                                                                                                                               Created0.0s iner doodarm-odoo-1                                                                                                                          Crea[+] Running 5/5odarm-db-1                                                                                                                            Star ✔ Volume "doodarm_filestore"                                                                                                                        Created0.0s e requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s ✔ Volume "doodarm_db"                                                                                                                               Created0.0s iner doodarm-odoo-1                                                                                                                          Crea[+] Running 5/5odarm-db-1                                                                                                                            Star ✔ Volume "doodarm_filestore"                                                                                                                        Created0.0s e requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s ✔ Volume "doodarm_db"                                                                                                                               Created0.0s iner doodarm-odoo-1                                                                                                                          Crea ✔ Container doodarm-db-1                                                                                                                            Started0.0s 
 ! db The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s 
 ✔ Container doodarm-odoo-1                                                                                                                          Started0.1s 


```

</details>

> so you can just do --build-arg WKTHTMLTOPDF_VERSION=version and it will use that no need to change the original file

Thanks, that will likely work. Well from usability point of view, personally I would prefer to set it by default to the arm supported version instead of throwing an error.

**Review by bosd — APPROVED** [review]
LGTM

**Comment by pedrobaeza** [comment]
Any news about how to publish this for merging it?

**Comment by bosd** [comment]
> Any news about how to publish this for merging it?

Not sure what you mean by this. 
My usability comment is non blocker. I'd say it is ready for merge.

**Comment by pedrobaeza** [comment]
It's not for you, @bosd. I mean this comment: https://github.com/Tecnativa/doodba/pull/573#issuecomment-1868580100

**Comment by bosd** [comment]
ping @PCatinean Can you attend to pedro's question?

**Comment by PCatinean** [comment]
@pedrobaeza here are the latest changes that add for support of multi-arch builds. The following adjustments have been made:

1. In order to not complicate the CI process by passing explicit arguments and also make the images work out of the box, version 14.0 and 15.0 get a "Forcing wkhtmltopdf version to 0.12.6.1" message for ARM64 builds (otherwise they wouldn't work anyway).
2. The `hooks` dir has been removed since the scripts inside (`push` & `build`) are deprecated. There are github actions that achieve the same result, have more support and also handle the docker buildx (required for building for other architectures).
3. The `platforms` variable is configurable per odoo version and can be extended with more architectures in the future.

The arm64 builds are a bit slow around this step `Building wheel for libsass (setup.py):`. I am sure it can be optimized but for the time being they are increasing the build time a bit.

Tests have been made locally and pushed to: https://hub.docker.com/repository/docker/pledra/doodba/tags?page=1&ordering=last_updated. The architectures available are shown there and when doing a regular pull it should reconcile with the architecture of the system automatically.

If there are any other tests I can make or info to share please let me know.

**Comment by bosd** [comment]
@pedrobaeza Is this one ok now? Ready for merge :pray:

**Comment by PCatinean** [comment]
@pedrobaeza since this is a bigger change to the workflow if you feel uncomfortable with it we can configure a different branch or a separate repo to push to inside this PR and test there before pushing to master.

**Review by pedrobaeza — APPROVED** [review]
Sorry, lack of time.

**Comment by pedrobaeza** [comment]
Please check when built if everything is OK and pushed, or tell me back if not.

**Pull request merged**